### PR TITLE
Bind/Unbind interpreters to note

### PR DESCRIPTION
**Pull request by Leemoonsoo**
Currently all interpreters are shared to all notes by default.
It'll be good If we can selectively bind/unbind available interpreters to note.

* [x] Ability to instantiate new interpreter and bind/unbind to note
* [x] API for bind/unbind interpreter
* [x] GUI Impelementation for bind/unbind interpreter
* [x] Save/load interpreter bindings
* [x] default property display
* [x] restore z.run() feature

### CODE REVIEW

**Comment by haebin** [comment]
Wouldn't it be better to use the name of interpreter setting as the interpreter selector as in '%<INTERPRETER>'?

Ex) %sqlLocal select blah blah ...

**Comment by Leemoonsoo** [comment]
@haebin Thanks for nice suggestion!

I think there're pros and cons.

**a. Current implementaion**

Cons - In a single notebook, can not use multiple instance of interpreter derived from the same type.
Pros - All zeppelin user will use the same %magic keyword for the same interpreter type. (therefore notebook can be run in any user's Zeppelin without %magic keyword modification)

**b. Haebin proposed way**
Pros - Can use multiple instance of interpreter derived from the same type. in a single note.
Cons - In case notebook is copied from other, to run, need to modify %magic keyword for local Zeppelin's setting name.

Guys what do you think? @haebin, @swkimme  any good idea?

One compromise is using always %spark but giving some arguments like
`%spark(name='spark-local')`

**Comment by swkimme** [comment]
I guess a.cons: 'using multiple instance of interpreter derived from the
same type in a single notebook' is  quite rare case.

I prefer current implementation for simplicity. (also it's more intuitive.)
supplying name as option seems to be quite good alternative like moon
sugested, %spark(name='spark-local')

**Comment by Leemoonsoo** [comment]
Now, i changed behavior to minimize the UX changes according to @swkimme 's suggestion.
Newly created note now have automatically selected interpreter sets.
(Existing note that created before this feature will still be asked)

Also it'll be asked before discard any changes on note setting menu.

Right buttons are now aligned
image

About description in interpreter menu, i was trying to make it look alike table that explain application's properties. Those kind of table we can see all around, hadoop, spark, etc. (eg. https://spark.apache.org/docs/1.1.0/configuration.html#application-properties)

**Comment by Leemoonsoo** [comment]
Summary of this PR

#### New features

* Creating / removing interpreter settings (interpreter instances)
* Binding/unbinding interpreter to notes

#### Improvements

* Align upper right icons in note page.

#### Bug fix

* Resolve netty dependency confliction
* Resolve NPE after Zeppelin restart when trying to to autocomplete
* Cancel paragraph now cancels only specific paragraph, not all other pending paragraphs.

#### API Changes

* Interpreter.interpret(String) -> Interpreter.interpret(String, InterpreterContext)
* Interpreter.cancel() -> Interpreter.cancel(InterpreterContext)
* Interpreter.getProgress() -> Interpreter.getProgress(InterpreterContext)


There're thing to be improved, like editing interpreter setting, use radio box instead of drag and drop to make default interpreter when binding interpreter, etc.
But all proposed features are working well and we can address improvements with separate issues.

Ready to merge

**Comment by Leemoonsoo** [comment]
About @haebin 's suggestion, i also agree on @swkimme that using the alternatives. `%spark(name='spark-local')` 
But i want to handle that in separate issue, we can discuss more about it on mailing list.

**Comment by Leemoonsoo** [comment]
If there're no further issues to discuss, i'm merging it.

**Comment by corneadoug** [comment]
Go for it, but we should also create issues for the suggestions so that we don't forget about it

**Comment by anthonycorbacho** [comment]
Just wait, There is some part in the rest api i want to talk about.

File: zeppelin-server/src/main/java/com/nflabs/zeppelin/rest/InterpreterRestApi.java [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Interpreter Rest API
+ *
+ */
+@Path("/interpreter")
+@Produces("application/json")
+public class InterpreterRestApi {
```
Comment (anthonycorbacho) [inline zeppelin-server/src/main/java/com/nflabs/zeppelin/rest/InterpreterRestApi.java]: Please add swagger annotation for the documentation

File: zeppelin-server/src/main/java/com/nflabs/zeppelin/rest/InterpreterRestApi.java [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+  public InterpreterRestApi(InterpreterFactory interpreterFactory) {
+    this.interpreterFactory = interpreterFactory;
+  }
+
+  /**
+   * List all interpreter settings
+     * @return
+   */
+  @GET
+  @Path("setting")
```
Comment (anthonycorbacho) [inline zeppelin-server/src/main/java/com/nflabs/zeppelin/rest/InterpreterRestApi.java]: by convention if you do a `get /ressource` you will get all list/collection of the resources
in this case, `get /interpreter ` is what you want to do, no need to do `get /interpreter/setting`

File: zeppelin-server/src/main/java/com/nflabs/zeppelin/rest/InterpreterRestApi.java [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+  }
+
+  /**
+   * Add new interpreter setting
+   * @param message
+   * @return
+   * @throws IOException
+   * @throws InterpreterException
+   */
+  @PUT
```
Comment (anthonycorbacho) [inline zeppelin-server/src/main/java/com/nflabs/zeppelin/rest/InterpreterRestApi.java]: If you want to create a new interpreter, use `POST`, `PUT` is use for update
And no need for `setting`, it should be `POST /interpreter`

File: zeppelin-server/src/main/java/com/nflabs/zeppelin/rest/InterpreterRestApi.java [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+  @DELETE
+  @Path("setting/{settingId}")
+  public Response removeSetting(@PathParam("settingId") String settingId) throws IOException {
+    logger.info("Remove interpreterSetting {}", settingId);
+    interpreterFactory.remove(settingId);
+    return new JsonResponse(Status.OK).build();
+  }
+
+  @GET
+  @Path("setting/restart/{settingId}")
```
Comment (anthonycorbacho) [inline zeppelin-server/src/main/java/com/nflabs/zeppelin/rest/InterpreterRestApi.java]: This is really confusing, `GET` should only get ressource, if you want to update a ressouce do `PUT`.

File: zeppelin-server/src/main/java/com/nflabs/zeppelin/rest/InterpreterRestApi.java [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+  public Response restartSetting(@PathParam("settingId") String settingId) {
+    logger.info("Restart interpreterSetting {}", settingId);
+    interpreterFactory.restart(settingId);
+    InterpreterSetting setting = interpreterFactory.get(settingId);
+    return new JsonResponse(Status.OK, "", setting).build();
```
Comment (anthonycorbacho) [inline zeppelin-server/src/main/java/com/nflabs/zeppelin/rest/InterpreterRestApi.java]: here, we should handle the fact that the settingId do not exist -> 404

File: zeppelin-server/src/main/java/com/nflabs/zeppelin/rest/InterpreterRestApi.java [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+  }
+
+  /**
+   * List all available interpreters by group
+   */
+  @GET
+  @Path("interpreter")
```
Comment (anthonycorbacho) [inline zeppelin-server/src/main/java/com/nflabs/zeppelin/rest/InterpreterRestApi.java]: no the to add twice interpreter, `GET interpreter/interpreter` is to wierd.

File: zeppelin-server/src/main/java/com/nflabs/zeppelin/rest/InterpreterRestApi.java [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+    interpreterFactory.add(request.getName(), request.getGroup(), p);
+    return new JsonResponse(Status.OK, "").build();
+  }
+
+  @DELETE
+  @Path("setting/{settingId}")
+  public Response removeSetting(@PathParam("settingId") String settingId) throws IOException {
+    logger.info("Remove interpreterSetting {}", settingId);
+    interpreterFactory.remove(settingId);
+    return new JsonResponse(Status.OK).build();
```
Comment (anthonycorbacho) [inline zeppelin-server/src/main/java/com/nflabs/zeppelin/rest/InterpreterRestApi.java]: what happen is something go wrong? we should be able to catch it and send a proper response like `NotFoundException` or  `BadRequestException`

File: zeppelin-server/src/main/java/com/nflabs/zeppelin/rest/InterpreterRestApi.java [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+   */
+  @PUT
+  @Path("setting")
+  public Response newSettings(String message) throws InterpreterException, IOException {
+    NewInterpreterSettingRequest request = gson.fromJson(message,
+        NewInterpreterSettingRequest.class);
+    Properties p = new Properties();
+    p.putAll(request.getProperties());
+    interpreterFactory.add(request.getName(), request.getGroup(), p);
+    return new JsonResponse(Status.OK, "").build();
```
Comment (anthonycorbacho) [inline zeppelin-server/src/main/java/com/nflabs/zeppelin/rest/InterpreterRestApi.java]: `PUT` responce for creation is not `200 OK` but `201 CREATED`.

**Comment by anthonycorbacho** [comment]
@Leemoonsoo where are the tests for the rest api ?

**Comment by bzz** [comment]
Great job @Leemoonsoo !
Looks good to me, except the conventions for REST API noted by @anthonycorbacho

**Comment by Leemoonsoo** [comment]
@anthonycorbacho I've changed REST API convention and added basic rest API test.
Ready to merge.

**Comment by anthonycorbacho** [comment]
@Leemoonsoo Looks better!!

**Pull request merged**